### PR TITLE
[issue] pcs support for multi opening locations

### PR DESCRIPTION
**Pull request by jorgeantonio21**
This PR addresses issue #187. Namely, the current logic contains refactoring of the following:

* It refactors the logic of method `fn open_multi_batches` in `ldt/src/ldt_based_pcs.rs`. The main change consists of the support for multi-point support for each prover data (leaf) matrix. That is to say, for each round and each matrix in each round's prover data allows for its own set of points for value openings.

* In the definition of associated type `OpenedValuesForRound<F>`, we exchanged the order of `OpenedValuesForMatrix<F>` and `OpenedValuesForPoint<F>`. This facilitates the underlying reasoning of the parametrization process. 

* A direct consequence of this refactor is the fact that in the `fn prove` method in `uni-stark/src/prover.rs`, we need to interchanged the indexes to have access to the right order openings.

### CODE REVIEW

File: ldt/src/ldt_based_pcs.rs [hunk 1]
```diff
@@ -211,18 +237,18 @@ where
             .zip_eq(values)
             .zip_eq(dims)
             .map(
-                |((points, opened_values_for_round_by_point), dims): (
-                    (&[EF], OpenedValuesForRound<EF>),
+                #[allow(clippy::type_complexity)]
```
Comment (jorgeantonio21) [inline ldt/src/ldt_based_pcs.rs:240]: Added after clippy warnings. Happy to remove this and the associated type.

File: ldt/src/ldt_based_pcs.rs [hunk 2]
```diff
@@ -250,11 +276,3 @@ where
             .verify(&quotient_mmcs, &dims, &commits, proof, challenger)
     }
 }
-
-fn transpose<T: Clone>(vec: Vec<Vec<T>>) -> Vec<Vec<T>> {
```
Comment (jorgeantonio21) [inline ldt/src/ldt_based_pcs.rs:254]: I think we can safely remove this code. Happy to added it back, if necessary.

File: ldt/src/ldt_based_pcs.rs [hunk 3]
```diff
@@ -128,32 +128,38 @@ where
...
                     .iter()
-                    .map(|(data, points)| {
+                    .map(|(data, points_per_matrix)| {
                         let matrices = self.mmcs.get_matrices(data);
-                        points
+
+                        matrices
                             .iter()
-                            .map(|&point| eval_at_point(&matrices, point))
+                            .enumerate()
```
Comment (dlubarov) [inline ldt/src/ldt_based_pcs.rs:158]: nit: could we zip with `points` to avoid needing indices?

Comment (dlubarov) [reply ldt/src/ldt_based_pcs.rs:158]: I think the other two uses of `enumerate()` could also potentially use `zip()`

File: ldt/src/ldt_based_pcs.rs (hunk 2 above)
Comment (dlubarov) [reply ldt/src/ldt_based_pcs.rs:254]: SGTM, let's remove for now

**Review by dlubarov — APPROVED** [review]
LGTM

File: ldt/src/ldt_based_pcs.rs (hunk 3 above)
Comment (jorgeantonio21) [reply ldt/src/ldt_based_pcs.rs:158]: Definitely ! Thanks for the suggestions

**Pull request merged**